### PR TITLE
Packages: Move element to packages maintained by Lerna

### PR DESCRIPTION
**Pull request by gziolo**
## Description

Related issues: #3955, #6594.
Follow up for: #6658.

This PR tries to provide Lerna setup similar to what we have in `WordPress/packages`. This would allow publishing source code of individual modules to npm to make them available for all plugin developers.

After moving `date` in #6658, this PR moves `element` to `packages folder maintained by `Lerna`.

## How has this been tested?
Manually:
* Removed `node_modules` folder.
* `npm install`
* `npm test`
* `npm run dev`
* `npm run build`
* `npm run package-plugin`

## Types of changes
Refactoring.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code has proper inline documentation.

### CODE REVIEW

File: packages/element/package.json [hunk 1]
```diff
@@ -0,0 +1,30 @@
+{
+	"name": "@wordpress/element",
+	"version": "0.0.1",
+	"description": "Element module for WordPress",
```
Comment (ntwb) [inline packages/element/package.json]: Wondering if `React` should be included in the title for < cough >~~SEO~~</ cough > marketing/discoverability purposes?

> `Element React module for WordPress`

**Review by ntwb — APPROVED** [review]
LGTM 👍 

p.s. Not a blocker, but see my comment on maybe changing the package title

**Comment by ntwb** [comment]
Possible merge conflict with #6758 when it comes time to merge

File: package.json [hunk 2]
```diff
@@ -20,7 +20,7 @@
 		"@wordpress/dom-ready": "1.0.4",
 		"@wordpress/hooks": "1.1.6",
 		"@wordpress/i18n": "1.1.0",
-		"@wordpress/is-shallow-equal": "1.0.1",
```
Comment (youknowriad) [inline package.json]: Any particular reason for this upgrade?

File: packages/element/package.json [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+	},
+	"bugs": {
+		"url": "https://github.com/WordPress/gutenberg/issues"
+	},
+	"main": "src/index.js",
+	"dependencies": {
+		"@wordpress/is-shallow-equal": "1.0.2",
+		"lodash": "4.17.5",
+		"react": "16.3.2",
+		"react-dom": "16.3.2"
```
Comment (youknowriad) [inline packages/element/package.json]: Does this mean, we can remove `react` and `react-dom` from the root level package.json?

File: package.json (hunk 2 above)
Comment (gziolo) [reply package.json]: Yes, there were some issues with the package outside of Gutenberg and it wasn't bumped in Gutenberg. See: https://github.com/WordPress/packages/pull/124.
We can do it separately.

File: packages/element/package.json (hunk 3 above)
Comment (gziolo) [reply packages/element/package.json]: Technically we can, but they won't be tracked in `package-json.lock`. In the previous PR we decided to keep them in sync with the root level `package.json`. We can revisit later. Personally, I would prefer to have it working with lock files and have only one place to include them.

File: package.json (hunk 2 above)
Comment (youknowriad) [reply package.json]: No, I'm fine keeping it, just curious

File: packages/element/package.json (hunk 3 above)
Comment (youknowriad) [reply packages/element/package.json]: I think it's not really important since it's just for test. But I agree it's a concern for other dependencies that are bundled and I think we should have a separate lock per module.

Comment (gziolo) [reply packages/element/package.json]: You are right, I forgot about it. As far as I remember, those local lock files, don't really work when using `lerna bootstrap`. @ntwb - or do they and we disabled to follow semantic versioning?

Comment (ntwb) [reply packages/element/package.json]: Hmmm, I've not got an accurate recollection, there was some trouble I think.

That said, https://github.com/WordPress/packages do not have lock files, why would the https://github.com/WordPress/gutenberg require lock files.

tl;dr: Lock files are for apps, not packages or modules

Comment (gziolo) [reply packages/element/package.json]: I found another option which should work, but it would create lots of copies of popular dependencies like `lodash`. See: https://github.com/lerna/lerna/issues/856#issuecomment-375838518:
`lerna exec -- npm i` - then every package would use its own lock file for conistency.

Comment (gziolo) [reply packages/element/package.json]: Maybe we should give it a try and have the lock file only for the root `package.json` because Gutenberg is an app 😄

Comment (youknowriad) [reply packages/element/package.json]: I understand your point but I think I disagree. The difference between "app" and "packages" is superficial. What's the app when your whole app is a collection of packages? How do you guarantee an npm install won't inadvertently break your application?

**Comment by gziolo** [comment]
Let's merge this one. I will open a follow-up to play with the lock files.

File: packages/element/package.json (hunk 3 above)
Comment (ntwb) [reply packages/element/package.json]: You can't which is one of my points in that if you look down an "app" or "package" with specific package versions using a lock file then what is the point of SemVer, you may as well release everything as a _major_ release and never worry about _patch_ or _minor_ releases for an "app" or "package" that is depended upon

Comment (youknowriad) [reply packages/element/package.json]: So you're ok with merging a PR and than your automatic CI job releasing your packages/app break because it go the wrong package you never tested?

I'm not. I see where you're coming from but that's true when dependencies get bundled together to build a single bundle which is not the case in our setup, each package is an app that generates its own script. We need predictability.

Comment (ntwb) [reply packages/element/package.json]: On a "package" side of things I am happy to let SemVer do what it was designed to do.

On an "app" I'm with you 100% and everything should be locked down.

Anything in between is tricky, which is where we are at here...

If the [Gutenberg packages](https://github.com/WordPress/gutenberg/tree/master/packages) are released using [SemVer](https://semver.org/) then using an explicit locked down version e.g. `1.0.0` and not `~1.0.0` or `^1.0.0` of that package for Gutenberg **with** a lock file will work perfectly for Gutenberg as the only time a dependancy is bumped is when it is manually bumped which would include testing that the version bump works as expected.

Individual packages are then free to be used in the wider community and will continue to adhere to SemVer allowing other packages to depend upon them and update as they see fit per their own SemVer policy.

Anyways, that's my take, should probably save this discussion for the follow up PR @gziolo has planned 😄

Comment (youknowriad) [reply packages/element/package.json]: But I'm still not sure I follow you. There's no difference between an "app" and a "package". All Gutenberg packages are "apps" because they're all released as such as WordPress scripts.

Gutenberg will never "require" or depend on Gutenberg packages because they're built in the same repository. That may change (not certain) when the merge happens but until then, all the packages are apps, there's no higher-level package-lock.json where they can be locked.

File: test/unit/setup-wp-aliases.js [hunk 4]
```diff
@@ -6,11 +6,9 @@ global.wp = {
...
 
-[
-	'element',
-	'blocks',
-].forEach( ( entryPointName ) => {
-	Object.defineProperty( global.wp, entryPointName, {
-		get: () => require( entryPointName ),
-	} );
+Object.defineProperty( global.wp, 'element', {
+	get: () => require( 'packages/element' ),
```
Comment (aduth) [inline test/unit/setup-wp-aliases.js]: How is it that `packages/element` refers to the `packages/element` folder, when it's not being defined as a path (and even if it were, required from `test/unit`)?

Comment (gziolo) [reply test/unit/setup-wp-aliases.js]: Jest resolves those paths starting from the root dir, that's why it works. 

Our goal is to get rid of this workaround completely when we have all modules moved to packages and we enforce using `createElement` in files using JSX - related commit https://github.com/WordPress/gutenberg/pull/6828/commits/69af0a483ea57f4831984b2c72dad9d4a56c904d from #6828.

The same applies to `wb.blocks` which is here temporarily until we remove deprecations which use this global explicitly.
